src/Simplify_Call.cpp

```diff
@@ -170,53 +171,49 @@ Expr Simplify::visit(const Call *op, ExprInfo *bounds) {
 
         const Type t = op->type;
 
-        uint64_t ub = 0;
-        int64_t sb = 0;
-        bool b_is_const_uint = const_uint(b, &ub);
-        bool b_is_const_int = const_int(b, &sb);
-        if (b_is_const_uint || b_is_const_int) {
-            if (b_is_const_int) {
-                ub = std::abs(sb);
-            }
+        // We might swap from a right to left shift or the reverse.
+        std::string result_op = op->name;
 
-            // Determine which direction to shift.
-            const bool b_is_pos = b_is_const_uint || (b_is_const_int && sb >= 0);
-            const bool b_is_neg = b_is_const_int && sb < 0;
-            const bool shift_left = ((op->is_intrinsic(Call::shift_left) && b_is_pos) ||
-                                     (op->is_intrinsic(Call::shift_right) && b_is_neg));
-            const bool shift_right = ((op->is_intrinsic(Call::shift_right) && b_is_pos) ||
-                                      (op->is_intrinsic(Call::shift_left) && b_is_neg));
+        // If we know the sign of this shift, change it to an unsigned shift.
+        if (b_info.min_defined && b_info.min >= 0) {
+            b = cast(b.type().with_code(halide_type_uint), b);
+        } else if (b_info.max_defined && b_info.max <= 0) {
+            result_op = Call::get_intrinsic_name(op->is_intrinsic(Call::shift_right) ? Call::shift_left : Call::shift_right);
+            b = mutate(cast(b.type().with_code(halide_type_uint), -b), nullptr);
+        }
 
+        // If the shift is by a constant, it should now be unsigned.
+        uint64_t ub = 0;
+        if (const_uint(b, &ub)) {
             // LLVM shl and shr instructions produce poison for
             // shifts >= typesize, so we will follow suit in our simplifier.
             if (ub >= (uint64_t)(t.bits())) {
                 return make_signed_integer_overflow(t);
```

Comment:
Should this be moved down? It looks like an unsigned `a` could trigger a signed_integer_overflow error

Reply:
I'm not sure. I think maybe "signed_integer_overflow" might be being used as a more general thing here. I think `unsigned <</>> t.bits()` is still meant to produce a poison value. Maybe `signed_integer_overflow` should be renamed? @abadams

Reply:
I think we should possibly just define what it does to be the arm behavior, so that the naming issue is moot.

Reply:
OK, but until then, the other parts of this PR are useful, so should we just leave this as-is?